cpp/src/arrow/compute/kernels/comparator.cc

```diff
@@ -0,0 +1,200 @@
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::GREATER_EQUAL>(ctx);
+    case CompareOperator::LOWER:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::LOWER>(ctx);
+    case CompareOperator::LOWER_EQUAL:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::LOWER_EQUAL>(ctx);
+  }
+
+  return nullptr;
+}
+
+std::shared_ptr<FilterFunction> MakeCompareFilterFunction(FunctionContext* ctx,
+                                                          const DataType& type,
+                                                          struct CompareOptions options) {
+  switch (type.id()) {
+    case UInt8Type::type_id:
+      return MakeCompareFilterFunctionType<UInt8Type>(ctx, options);
+    case Int8Type::type_id:
+      return MakeCompareFilterFunctionType<Int8Type>(ctx, options);
+    case UInt16Type::type_id:
+      return MakeCompareFilterFunctionType<UInt16Type>(ctx, options);
+    case Int16Type::type_id:
+      return MakeCompareFilterFunctionType<Int16Type>(ctx, options);
+    case UInt32Type::type_id:
+      return MakeCompareFilterFunctionType<UInt32Type>(ctx, options);
+    case Int32Type::type_id:
+      return MakeCompareFilterFunctionType<Int32Type>(ctx, options);
+    case UInt64Type::type_id:
+      return MakeCompareFilterFunctionType<UInt64Type>(ctx, options);
+    case Int64Type::type_id:
+      return MakeCompareFilterFunctionType<Int64Type>(ctx, options);
+    case FloatType::type_id:
+      return MakeCompareFilterFunctionType<FloatType>(ctx, options);
+    case DoubleType::type_id:
+      return MakeCompareFilterFunctionType<DoubleType>(ctx, options);
+    case Date32Type::type_id:
+      return MakeCompareFilterFunctionType<Date32Type>(ctx, options);
+    case Date64Type::type_id:
+      return MakeCompareFilterFunctionType<Date64Type>(ctx, options);
+    case TimestampType::type_id:
+      return MakeCompareFilterFunctionType<TimestampType>(ctx, options);
+    case Time32Type::type_id:
+      return MakeCompareFilterFunctionType<Time32Type>(ctx, options);
+    case Time64Type::type_id:
+      return MakeCompareFilterFunctionType<Time64Type>(ctx, options);
+    default:
+      return nullptr;
+  }
+}
+
+ARROW_EXPORT
+Status Compare(FunctionContext* context, const Datum& left, const Datum& right,
+               struct CompareOptions options, Datum* out) {
+  DCHECK(out);
+
+  if (!left.is_array())
+    return Status::Invalid("Compare expects an Array as left operand");
+  if (!right.is_scalar())
+    return Status::Invalid("Compare expects an Scalar as right operand");
+  if (!left.type()->Equals(right.type()))
+    return Status::Invalid("Compare expects left and right of same type");
+
+  auto array = left.make_array();
+  auto scalar = right.scalar();
+  auto type = array->type();
+  auto fn = MakeCompareFilterFunction(context, *type, options);
+
+  if (!fn) return Status::Invalid("Compare not implemented for type ", type->ToString());
```

Comment:
nit: I don't know if we have a standard and its not covered by the style guide, but consider fn == nullptr (or don't).